connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java

```diff
@@ -168,22 +170,26 @@ public synchronized <V> V execute(Operation<V> operation, Stage stage, Class<?>
      * @throws Exception rethrow if a non-retriable Exception is thrown by the operation
      */
     protected <V> V execAndRetry(Operation<V> operation) throws Exception {
+        RetriableException lastException = null;
         int attempt = 0;
         long startTime = time.milliseconds();
         long deadline = (errorRetryTimeout >= 0) ? startTime + errorRetryTimeout : Long.MAX_VALUE;
         do {
+            if (stopping) {
+                log.trace("Shutdown has been scheduled. Marking operation as failed.");
+                context.error(new ConnectException(lastException));
+                return null;
+            }
```

Comment:
It'd be cleaner if we put this inside the `catch` block, so that we wouldn't have to introduce the possibly-null `lastException` field.
Thinking something like this:
```java
do {
    try {
        attempt++;
        return operation.call();
    } catch (RetriableException e) {
        log.trace("Caught a retriable exception while executing {} operation with {}", context.stage(), context.executingClass());
        errorHandlingMetrics.recordFailure();
        if (time.milliseconds() < deadline) {
            backoff(attempt, deadline);
            errorHandlingMetrics.recordRetry();
        } else {
            log.trace("Can't retry. start={}, attempt={}, deadline={}", startTime, attempt, deadline);
            context.error(e);
            return null;
        }
        if (stopping) {
            log.trace("Shutdown has been scheduled. Marking operation as failed.");
            context.error(e);
            return null;
        }
    } finally {
        context.attempt(attempt);
    }
} while (true);
```

Reply:
Fair enough, I guess we don't need to worry too much about getting stopped before the first attempt.